src/gleam/erlang/os.gleam

```diff
@@ -63,3 +63,48 @@ pub external fn set_env(name: String, value: String) -> Nil =
 ///
 pub external fn unset_env(name: String) -> Nil =
   "gleam_erlang_ffi" "unset_env"
+
+/// Run a command in a command shell.
+///
+/// This function returns the output of the command as a String.
+///
+/// ## Examples
+///
+///    > cmd("ls")
+///    "Code\nDesktop\nDocuments\nMusic\nPictures\nVideos"
+///
+///    > cmd("ls -a")
+///    ".\n..\n.asdf\n.config\nCode\nDesktop\nDocuments\nMusic\nPictures\nVideos"
+///
+pub external fn cmd(String) -> String =
+  "gleam_erlang_ffi" "cmd"
+
+/// Represents operating system kernels
+pub type Kernel {
+  // A popular open source operating system kernel
+  Linux
+  // The kernel used by Apple operating systems, such as macOS and iOS.
+  Darwin
+  // The kernel underlying the various editions of the Windows operating system.
+  Nt
+  // An operating system kernel other than Linux, Darwin, or NT.
+  Other(String)
+}
+
+/// Returns the kernel of the host operating system.
+///
+/// Unknown kernels are reported as `Other(String)`; e.g. `Other("freebsd")`.
+///
+/// ## Examples
+///
+///    > kernel()
+///    Linux
+///    > kernel()
+///    Darwin
+///    > kernel()
+///    Nt
+///    > kernel()
+///    Other("freebsd")
+///
+pub external fn kernel() -> Kernel =
```

Comment:
Would it be clearer to report the operating system than the kernel?

```rust
pub fn which() -> OperatingSystem =
  "gleam_erlang_ffi" "kernel"

pub type OperatingSystem {
    Linux
    Macos
    Ios
    Freebsd
    Dragonfly
    Netbsd
    Openbsd
    Solaris
    Android
    Windows
    Other(String)
}
```

Reply:
The impetus for using `kernel` largely stems from Linux and Windows/NT. Linux is a kernel that undergirds a _huge number_ of operating systems such as Ubuntu, Fedora, RHEL, Arch, etc. Similarly, "Windows" is a family of operating systems built upon the NT kernel.

All this to say, the list would look more like this:
```rust
pub type OperatingSystem {
  // Apple / Darwin
  Macos
  Tvos
  Watchos
  Ios
  Ipados // Separate from iOS
  // Windows Family
  Windows11
  Windows10
  Windows81
  Windows8
  // ... etc
  // Linux based
  Fedora
  Rhel
  Centos
  Coreos
  Ubuntu
  Debian
  // etc
  // BSD
  Freebsd
  Dragonflybsd
  Ghostbsd
  Netbsd
  Openbsd
  // ... etc
}
```

FWIW, too, [`os:type()`](https://www.erlang.org/doc/man/os.html#type-0) does not report the operating system name, only it's "family" (actually only `unix` or `win32`) and its kernel (misleadingly named `Osname`, but on macOS it returns `darwin` and on Fedora it returns `linux` :shrug:). If we wanted to supply the actual name of the operating system, we'd have to write the trickery to tease this information out of the host.

Reply:
I don't think we need to be that fine grained, the above list comes from the Rust docs and it seems fairly sensible.

Having said that kernel is probably good enough. Let's add FreeBSD though as we do have FreeBSD users.

Reply:
After doing even more research, I've come away with a new perspective.

We could probably trade `Nt` for `Windows` or `WindowsNt`. "Windows NT" might actually be the name of the kernel anyways; the kernel may either share the OS family name or may not actually have its own name. `Windows` is a little bit vague because there were Windows operating systems that did not use the [Windows] NT kernel[^1], though these are all so old that we probably don't need to care.

It turns out that [Apple says that Darwin is an operating system](https://github.com/apple/darwin-xnu), using the XNU kernel. `os:type()` reports macOS' `Osname` as `Darwin` which is truer than I gave it credit for. However, I believe the only Darwin-based system that has an official BEAM implementation is macOS, so `Darwin` and `Macos` are, for the time being, interchangeable.

> Solaris

I think this falls in the camp of "too old, doesn't need to be first-class" unless you feel differently.

> Android

AFAIK, there's no official BEAM implementation for Android, so I'd lean towards not making it first-class either. Happy to add this if that ever changes though! :grin: 

---

**TL:DR**: Given these revelations it might be time to eschew the `kernel` name and replace it with something like `family` wherein we can say `Linux` is a family, `Windows` (or `WindowsNt`, more precisely) is a family, and `Darwin` or `Macos` is a family.

Off-hand options:
1. `Linux`, `WindowsNT`, `Darwin`, `FreeBsd`, `Other(String)` ("the strict option")
2. `Linux`, `Windows`, `Macos`, `FreeBsd`, `Other(String)` ("the 🤷 close enough option")

@lpil What do you think?

[^1]: Windows NT 3.1, a member of the Windows NT family, is not to be confused with Windows 3.1, which is not. Windows 3.1 was released a year prior using a different (monolithic DOS-based as opposed to micro Mach-based) kernel. Microsoft also created another kernel named "Windows 9x" for the Windows 95, 98 and ME operating systems that was also monolithic and DOS-based.

Reply:
Haha wow, I feel you've been on quite the knowledge rabbithole. TIL there's a Windows NT 3.1.

I like the family idea a lot. Feels like it strikes a good balance between precision and hand-waveyness to make things fit nicely. My only note is that `WindowsNT` should be `WindowsNt` as we snake_case names in the generated code and we don't want it to be `windows_n_t`.